### ISSUE TITLE
Netlify Form Detection

### ISSUE THREAD

**Issue by shootyourbricks** [webstudio-is/webstudio#4529]
Netlify automatically detects html forms during deployment. WebStudio forms (and html forms in an embed component) are not detected unless the entry-server-routing.js file is removed from the static build.

Steps to reproduce bug:
1. Create a form in WebStudio (can be form component or html embed). Form must have data-netlify="true" as explained in [Netlify's docs ](https://docs.netlify.com/forms/setup/?_gl=1%2azz4kjm%2a_gcl_au%2aMTk3NDc1MTcyMC4xNzMzMTYyODIxLjMzNDEyNzM3Mi4xNzMzMzMwNzE0LjE3MzMzMzA3NTU.)
2. Deploy dynamic site via CLI or static site via Github or File upload to Netlify
3. Check Netlify Forms UI or try submitting a form response. The form will not submit or show a response because Netlify didn't detect it

Steps to get the form detected (NOT an ideal workaround)
1. Create a form in WebStudio
2. Export static files
3. Delete entry-server-routing.js file
4. Deploy to netlify.

**Comment by shootyourbricks** [comment]
How important is the server-routing.js script? Would it be bad to remove it from one page as a temporary work-around?

**Comment by JayaKrishnaNamburu** [comment]
Hey @shootyourbricks i started looking it today. Will share more details tomorrow 👍🏽

**Comment by JayaKrishnaNamburu** [comment]
Here is a detailed analysis on why the netlify form detection is not working right now. It is because of `@remix-run/react` and `react-router-dom`. And the `form` that is being used on the canvas is a wrapper around the `Form` from the `@remix-run/react` package. 

The webhook-form that is available is not the best bet for the use-case. The other workaround is try using code-embed, but need some more testing on the solid use-case

Here is a total analysis of the use-case, i opened an issue on netlify. Because even they need to find a solution or else their form detection won't work out of the box anymore.
https://github.com/netlify/cli/issues/6955

cc @kof

**Comment by kof** [comment]
In other words netlify needs to fix their hack, otherwise user needs to use a regular Form component, not a Webhook Form, which uses Remix Form and Netlify breaks it. With a regular form user can't use webstudio form handling logic.

**Comment by shootyourbricks** [comment]
Here's the thing though. I'm not using a webhook form. I even tried embedding a form via html component and that didn't work either

**Comment by kof** [comment]
@shootyourbricks in what sense it didn't work? regular form element doesn't have functionality beyond submitting regular html ...

**Comment by shootyourbricks** [comment]
@kof Netlify didn't detect the form unless that js script was removed

**Comment by shootyourbricks** [comment]
I'm trying to test netlify ssg, however I have not been successful as I am running into an unrelated error. Posted a help post on Discord. Will share my results as soon as I have it figured out.

**Comment by JayaKrishnaNamburu** [comment]
Ok, i have spent some good amount of time in the last couple of days in this. The actual forms detection behind the `netlify.com` is bit of a black box. So, it took multiple trail and errors to get to this conclusion.

- Netlify can't detect forms if it's deployed as a remix project.
- Remix adds `action` to the form, and probably there need to be a `action` handler similar to webhook form. And seems like webhook form is the only way going forward.
- The third trail which i got to work is, do a form in webstudio. And then build / export the project as `ssg-netlify` template. And when you deploy this, only then netlify is able to detect. And again, probably this might not be the good us they have. First you need to deploy this `ssg-netlify` site and then go to `forms` and then enable form detection. Next you need to redeploy again for the form detection to kick in. And then it's starts tracking. Basically you need to deploy two times, and switch on the detection in between these two steps.

<img width="1800" alt="Screenshot 2024-12-12 at 8 26 05 PM" src="https://github.com/user-attachments/assets/3f9a9768-1a89-408e-a4c2-0086b61b9fb0" />

<img width="1375" alt="Screenshot 2024-12-12 at 8 51 12 PM" src="https://github.com/user-attachments/assets/8b5a984e-23f0-4267-a365-dc3161f41481" />

**Comment by shootyourbricks** [comment]
Thanks for working on this. 

> And seems like webhook form is the only way going forward.

I don't quite understand what you're saying about the webhook forms. Are you saying it's better to use the webhook form? Which did you use in the example? Netlify works with actionless forms in plain HTML files (not ws). 


I successfully got Netlify to detect a regular form with the SSG and two deploys as you explained but the form wouldn't submit. It would go to a 404 when submitting.

**Comment by JayaKrishnaNamburu** [comment]
> I don't quite understand what you're saying about the webhook forms. Are you saying it's better to use the webhook form? Which did you use in the example? Netlify works with actionless forms in plain HTML files (not ws).

If you are looking to deploy the app as a `remix` application. Then it is webhook form.

> I successfully got Netlify to detect a regular form with the SSG and two deploys as you explained but the form wouldn't submit. It would go to a 404 when submitting.

Check if you are redirection to some page under the `action` value in the from. If you are uploading plain files from the `ssg-netlify` approach then it should work. If not, share your project link, i can check once.

**Comment by shootyourbricks** [comment]
@JayaKrishnaNamburu There is no action for the form. In fact it looks like even though Netlify detected the form, it hasn't changed any of the actual attributes. "data-netlify" is still there and the hidden input is missing.

Here's the link for the webstudio project: https://p-bc461e1d-1866-4bbc-b54f-dae4c136fd5b.apps.webstudio.is/?authToken=e5fd87ef-8ba2-4a56-93d3-2a5bdae6e1cf&mode=preview

Here's the link to the deployed site on Netlify: https://ws-form-test.netlify.app/

**Comment by shootyourbricks** [comment]
@JayaKrishnaNamburu Are you able to take a look at my project?

**Comment by shootyourbricks** [comment]
I got it working by adding a hidden input with the name value of "form-name" and the value as the actual form name (in this case "contact" and "WSNative") This is what Netlify was supposed to add during form detection. Did you have to add that manually? @JayaKrishnaNamburu

**Comment by shootyourbricks** [comment]
It looks like if I remove the entry-server-routing.js script the deployment works as it should (I remove it after running the netlify build command but before running netlify deploy)

Why is this script being created when using the ssg template? @JayaKrishnaNamburu

**Comment by shootyourbricks** [comment]
Looks like solving the issue with Remix apps isn't something that Netlify plans to do. (A support engineer from Netlify said "There were “talks” to improve this, but I won’t expect that to land any time soon, if at all.")

It can be solved through a work around as explained here: https://opennext.js.org/netlify/forms#workaround-for-netlify-forms

Is this something that WebStudio could implement in order to have compatibility with Netlify forms (if so, should I move this to Github discussions, or leave it here in issues?)

**Comment by JayaKrishnaNamburu** [comment]
Looks like, netlify don't work with auto detect when used along with any framework. Which they shared it in their GitHub issue here
https://github.com/netlify/cli/issues/6955#issuecomment-2567121291

So, there is a workaround which is shared by their team
https://opennext.js.org/netlify/forms#workaround-for-netlify-forms

Which is not possible to implement using the builder at `webstudio.is` if you want to follow the auto detection route for forms. I built an example by following the steps for `form-name` and `input-name`. But netlify still doesn't seem to detect them in the `netlify-functions` neither in `ssg-netlify` templates.

And regarding to your question on
>  It seems that the entry-server-routing.js script that blocks form detection/processing

Currently when you select a `netlify` template. It generates a `entry.server.tsx` file. And `ssg` don't generate that too. But i am not sure why you have `entry-server-routing.js` file. Share the template name that you are using.


I am not sure, if the workaround that netlify team is suggesting can be built right into the wwebstudio build system. More in this @kof can help if we ever decide to bring in support for that workaround. Because tomorrow the team might come up with new workaround and the builder can't implement and break existing ones.

Bottom line, netlify form detection doesn't work with anything other than plain html files.
https://answers.netlify.com/t/netlify-cli-cant-detect-forms-from-when-deploying-remix-app/135876

**Comment by shootyourbricks** [comment]
> Currently when you select a `netlify` template. It generates a `entry.server.tsx` file. And `ssg` don't generate that too. But i am not sure why you have `entry-server-routing.js` file. Share the template name that you are using.

Whether I use the netlify-ssg template or simply export and download static files for vanilla or netllify builds, I always get the js script

Here's my project https://p-3e6aec26-235a-41d3-a79b-6dc9a6e8f287.apps.webstudio.is/?authToken=37314ef1-5936-406d-9d7a-729638817506&mode=preview

Attached is the index.html file (renamed to .txt) and you can see the script at the bottom. /assets/entries/entry-server-routing.CcTuSK9F.js That's the one that keeps the form detection from fully working
[index.txt](https://github.com/user-attachments/files/18335775/index.txt)

**Comment by JayaKrishnaNamburu** [comment]
Hey, i still haven't gotten any update from them. So, opened a discussion on the vike repo. To get some opinions on it
https://github.com/vikejs/vike/discussions/2056

But honestly, all these tools, `vike`, `webstudio` etc are working as expected. It's just that netlify's form detection and auto form handling doesn't work out of the box with any js based frameworks. They just work out of the box with pure `.html` files.

**Comment by shootyourbricks** [comment]
> To get some opinions on it [vikejs/vike#2056](https://github.com/vikejs/vike/discussions/2056)

Thanks. I'll probably end up using a CMS or doing a workaround so it's not a huge issue. Anyway just FYI, I can get the forms working, so I don't think vike is really messing up the forms, it just seems like vike is preventing Netlify from modifying the html docs (adding the hidden input and removing the form attribute)

**Comment by shootyourbricks** [comment]
@kof You may close this issue as it is not technically a "bug" (but rather an incompatibility with netlify forms and js frameworks) and I have provided workaround details in the original post.
